File: vermyth/grimoire/repositories/divergence.py
from __future__ import annotations

import sqlite3
from datetime import datetime, timezone
from typing import Any

from vermyth.schema import (
    DivergenceReport,
    DivergenceStatus,
    DivergenceThresholds,
    DivergenceThresholds_DEFAULT,
)


class DivergenceRepository:
    """Repository methods for divergence persistence."""

    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn

    def write_divergence_report(self, report: DivergenceReport) -> None:
        try:
            cur = self._conn.cursor()
            cur.execute(
                """
                INSERT OR REPLACE INTO divergence_reports (
                    cast_id, parent_cast_id, l2_magnitude, cosine_distance, status, computed_at
                ) VALUES (?, ?, ?, ?, ?, ?)
                """,
                (
                    report.cast_id,
                    report.parent_cast_id,
                    float(report.l2_magnitude),
                    float(report.cosine_distance),
                    report.status.name,
                    report.computed_at.isoformat(),
                ),
            )
            self._conn.commit()
        except sqlite3.Error as exc:
            raise RuntimeError(str(exc)) from exc

    def read_divergence_report(self, cast_id: str) -> DivergenceReport:
        try:
            cur = self._conn.cursor()
            cur.execute("SELECT * FROM divergence_reports WHERE cast_id = ?", (cast_id,))
            row = cur.fetchone()
            if row is None:
                raise KeyError(cast_id)
            return DivergenceReport(
                cast_id=row["cast_id"],
                parent_cast_id=row["parent_cast_id"],
                l2_magnitude=float(row["l2_magnitude"]),
                cosine_distance=float(row["cosine_distance"]),
                status=DivergenceStatus[row["status"]],
                computed_at=datetime.fromisoformat(row["computed_at"]),
            )
        except KeyError:
            raise
        except sqlite3.Error as exc:
            raise RuntimeError(str(exc)) from exc

    def query_divergence_reports(
        self,
        status: DivergenceStatus | None = None,
        limit: int = 50,
        *,
        since: datetime | None = None,
    ) -> list[DivergenceReport]:
        try:
            cur = self._conn.cursor()
            where: list[str] = []
            params: list[object] = []
            if status is not None:
                where.append("status = ?")
                params.append(status.name)
            if since is not None:
                where.append("computed_at >= ?")
                params.append(since.isoformat())
            where_sql = ""
            if where:
                where_sql = "WHERE " + " AND ".join(where)
            params.append(int(limit))
            cur.execute(
                f"SELECT * FROM divergence_reports {where_sql} ORDER BY computed_at DESC LIMIT ?",
                tuple(params),
            )
            out: list[DivergenceReport] = []
            for row in cur.fetchall():
                out.append(
                    DivergenceReport(
                        cast_id=row["cast_id"],
                        parent_cast_id=row["parent_cast_id"],
                        l2_magnitude=float(row["l2_magnitude"]),
                        cosine_distance=float(row["cosine_distance"]),
                        status=DivergenceStatus[row["status"]],
                        computed_at=datetime.fromisoformat(row["computed_at"]),
                    )
                )
            return out
        except sqlite3.Error as exc:
            raise RuntimeError(str(exc)) from exc

    def write_divergence_thresholds(self, thresholds: DivergenceThresholds) -> None:
        try:
            cur = self._conn.cursor()
            cur.execute(
                """
                INSERT OR REPLACE INTO divergence_thresholds (
                    id, l2_stable_max, l2_diverged_min, cosine_stable_max, cosine_diverged_min, updated_at
                ) VALUES (1, ?, ?, ?, ?, ?)
                """,
                (
                    float(thresholds.l2_stable_max),
                    float(thresholds.l2_diverged_min),
                    float(thresholds.cosine_stable_max),
                    float(thresholds.cosine_diverged_min),
                    datetime.now(timezone.utc).isoformat(),
                ),
            )
            self._conn.commit()
        except sqlite3.Error as exc:
            raise RuntimeError(str(exc)) from exc

    def read_divergence_thresholds(self) -> DivergenceThresholds:
        try:
            cur = self._conn.cursor()
            cur.execute("SELECT * FROM divergence_thresholds WHERE id = 1")
            row = cur.fetchone()
            if row is None:
                return DivergenceThresholds_DEFAULT
            return DivergenceThresholds(
                l2_stable_max=float(row["l2_stable_max"]),
                l2_diverged_min=float(row["l2_diverged_min"]),
                cosine_stable_max=float(row["cosine_stable_max"]),
                cosine_diverged_min=float(row["cosine_diverged_min"]),
            )
        except sqlite3.Error as exc:
            raise RuntimeError(str(exc)) from exc

    def drift_branches(self, limit: int = 25) -> list[dict[str, Any]]:
        try:
            cur = self._conn.cursor()
            cur.execute(
                """
                SELECT
                    cr.branch_id AS branch_id,
                    COUNT(*) AS casts_with_divergence,
                    SUM(CASE WHEN dr.status = 'DIVERGED' THEN 1 ELSE 0 END) AS diverged_count,
                    SUM(CASE WHEN dr.status = 'DRIFTING' THEN 1 ELSE 0 END) AS drifting_count,
                    MAX(dr.l2_magnitude) AS max_l2,
                    MAX(dr.cosine_distance) AS max_cosine_distance,
                    MAX(dr.computed_at) AS latest_computed_at
                FROM divergence_reports dr
                JOIN cast_results cr ON cr.cast_id = dr.cast_id
                WHERE cr.branch_id IS NOT NULL
                GROUP BY cr.branch_id
                ORDER BY
                    diverged_count DESC,
                    drifting_count DESC,
                    max_l2 DESC,
                    max_cosine_distance DESC,
                    latest_computed_at DESC
                LIMIT ?
                """,
                (int(limit),),
            )
            out: list[dict[str, Any]] = []
            for row in cur.fetchall():
                out.append(
                    {
                        "branch_id": row["branch_id"],
                        "casts_with_divergence": int(row["casts_with_divergence"] or 0),
                        "diverged_count": int(row["diverged_count"] or 0),
                        "drifting_count": int(row["drifting_count"] or 0),
                        "max_l2": float(row["max_l2"] or 0.0),
                        "max_cosine_distance": float(row["max_cosine_distance"] or 0.0),
                        "latest_computed_at": row["latest_computed_at"],
                    }
                )
            return out
        except sqlite3.Error as exc:
            raise RuntimeError(str(exc)) from exc

    def cast_pairs_missing_divergence_reports(
        self, limit: int = 500
    ) -> list[tuple[str, str]]:
        try:
            cur = self._conn.cursor()
            cur.execute(
                """
                SELECT
                    cr.cast_id AS cast_id,
                    json_extract(cr.lineage_json, '$.parent_cast_id') AS parent_cast_id
                FROM cast_results cr
                LEFT JOIN divergence_reports dr ON dr.cast_id = cr.cast_id
                WHERE cr.lineage_json IS NOT NULL
                  AND dr.cast_id IS NULL
                ORDER BY cr.timestamp DESC
                LIMIT ?
                """,
                (int(limit),),
            )
            out: list[tuple[str, str]] = []
            for row in cur.fetchall():
                cid = str(row["cast_id"])
                pid = row["parent_cast_id"]
                if pid is None:
                    continue
                out.append((cid, str(pid)))
            return out
        except sqlite3.Error as exc:
            raise RuntimeError(str(exc)) from exc

File: typescript/src/grimoire/repositories/divergence.ts
import BetterSqlite3 from "better-sqlite3";

import {
  defaultDivergenceThresholds,
  type DivergenceReport,
  type DivergenceThresholds,
} from "../../schema/divergence.js";
import type { DivergenceStatus } from "../../schema/enums.js";

type SqliteDb = InstanceType<typeof BetterSqlite3>;

export class DivergenceRepository {
  constructor(private readonly db: SqliteDb) {}

  readDivergenceThresholds(): DivergenceThresholds {
    const row = this.db
      .prepare("SELECT * FROM divergence_thresholds WHERE id = 1")
      .get() as
      | {
          l2_stable_max: number;
          l2_diverged_min: number;
          cosine_stable_max: number;
          cosine_diverged_min: number;
        }
      | undefined;
    if (!row) {
      return defaultDivergenceThresholds();
    }
    return {
      l2StableMax: Number(row.l2_stable_max),
      l2DivergedMin: Number(row.l2_diverged_min),
      cosineStableMax: Number(row.cosine_stable_max),
      cosineDivergedMin: Number(row.cosine_diverged_min),
    };
  }

  readDivergenceReport(castId: string): DivergenceReport | null {
    const row = this.db.prepare("SELECT * FROM divergence_reports WHERE cast_id = ?").get(castId) as
      | {
          cast_id: string;
          parent_cast_id: string;
          l2_magnitude: number;
          cosine_distance: number;
          status: string;
          computed_at: string;
        }
      | undefined;
    if (!row) {
      return null;
    }
    return {
      castId: String(row.cast_id),
      parentCastId: String(row.parent_cast_id),
      l2Magnitude: Number(row.l2_magnitude),
      cosineDistance: Number(row.cosine_distance),
      status: row.status as DivergenceStatus,
      computedAt: new Date(String(row.computed_at)),
      basisNote: null,
    };
  }

  driftBranches(limit: number): Array<Record<string, unknown>> {
    const rows = this.db
      .prepare(
        `
      SELECT
        cr.branch_id AS branch_id,
        COUNT(*) AS casts_with_divergence,
        SUM(CASE WHEN dr.status = 'DIVERGED' THEN 1 ELSE 0 END) AS diverged_count,
        SUM(CASE WHEN dr.status = 'DRIFTING' THEN 1 ELSE 0 END) AS drifting_count,
        MAX(dr.l2_magnitude) AS max_l2,
        MAX(dr.cosine_distance) AS max_cosine_distance,
        MAX(dr.computed_at) AS latest_computed_at
      FROM divergence_reports dr
      JOIN cast_results cr ON cr.cast_id = dr.cast_id
      WHERE cr.branch_id IS NOT NULL
      GROUP BY cr.branch_id
      ORDER BY
        diverged_count DESC,
        drifting_count DESC,
        max_l2 DESC,
        max_cosine_distance DESC,
        latest_computed_at DESC
      LIMIT ?
    `,
      )
      .all(Math.max(1, Math.min(500, Math.floor(limit)))) as Array<{
      branch_id: string;
      casts_with_divergence: number;
      diverged_count: number;
      drifting_count: number;
      max_l2: number;
      max_cosine_distance: number;
      latest_computed_at: string;
    }>;
    return rows.map((row) => ({
      branch_id: row.branch_id,
      casts_with_divergence: Number(row.casts_with_divergence ?? 0),
      diverged_count: Number(row.diverged_count ?? 0),
      drifting_count: Number(row.drifting_count ?? 0),
      max_l2: Number(row.max_l2 ?? 0),
      max_cosine_distance: Number(row.max_cosine_distance ?? 0),
      latest_computed_at: row.latest_computed_at,
    }));
  }
}
